src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs

```diff
@@ -1499,22 +1505,31 @@ private MethodSymbol FindEntryPoint(CancellationToken cancellationToken, out Imm
                         diagnostics.Add(ErrorCode.ERR_NoMainInClass, mainType.Locations.First(), mainType);
                     }
                 }
-                else if (viableEntryPoints.Count > 1)
+                else if (nonAsyncCount == 1)
+                {
+                    entryPoint = viableEntryPoints.Single(c => !c.IsAsync);
+                }
+                else if (nonAsyncCount == 0 && asyncCount == 1)
+                {
+                    diagnostics.AddRange(possibleAsyncMainDiagnostics);
+                    entryPoint = viableEntryPoints.Single();
+                }
+                else if (nonAsyncCount == 0 && !asyncOk)
+                {
+                    diagnostics.AddRange(possibleAsyncMainDiagnostics);
```

Comment:
So I went and did some more testing, and apparently all of this complex logic isn't actually very useful.  In shipping csc, there is no way to have an async method that meets the entrypoint requirements without compiler errors. 

So all of this stuff that was supposed to be for backcompat is pretty useless...